File: integrations/kotlin/src/jsMain/kotlin/angular/ts/Directive.kt
package angular.ts

import angular.ts.generated.Scope as RawScope
import org.w3c.dom.HTMLElement

public enum class DirectiveRestrict(
    internal val raw: String,
) {
    Attribute("A"),
    Element("E"),
    AttributeElement("AE"),
    ElementAttribute("EA"),
}

public typealias DirectiveLink<TController> = (
    scope: Scope<Any>,
    element: HTMLElement,
    controller: TController?,
    transclude: dynamic,
) -> Unit

public data class Directive<TController> public constructor(
    public val template: String? = null,
    public val templateUrl: String? = null,
    public val controller: InjectableFactory<TController>? = null,
    public val controllerAs: String? = null,
    public val restrict: DirectiveRestrict = DirectiveRestrict.ElementAttribute,
    public val scope: Boolean = false,
    public val bindToController: Boolean = false,
    public val require: Any? = null,
    public val link: DirectiveLink<TController>? = null,
    public val preLink: DirectiveLink<TController>? = null,
    public val postLink: DirectiveLink<TController>? = null,
)

internal fun Directive<*>.toJs(): dynamic {
    val raw = js("{}")

    raw.restrict = restrict.raw
    raw.scope = scope
    raw.bindToController = bindToController

    if (template != null) raw.template = template
    if (templateUrl != null) raw.templateUrl = templateUrl
    if (controller != null) raw.controller = controller.toJs()
    if (controllerAs != null) raw.controllerAs = controllerAs
    if (require != null) raw.require = require
    if (link != null) raw.link = link.toJs(require != null)
    if (preLink != null || postLink != null) {
        raw.link = prePostLinksToJs(preLink, postLink, require != null)
    }

    return raw
}

private fun <TController> DirectiveLink<TController>.toJs(hasRequire: Boolean): dynamic =
    { scope: dynamic, element: dynamic, third: dynamic, fourth: dynamic ->
        val controller = if (hasRequire) third else null
        val transclude = if (hasRequire) fourth else third

        this(
            Scope(scope.unsafeCast<RawScope>()),
            element.unsafeCast<HTMLElement>(),
            controller.unsafeCast<TController?>(),
            transclude,
        )
    }

private fun <TController> prePostLinksToJs(
    preLink: DirectiveLink<TController>?,
    postLink: DirectiveLink<TController>?,
    hasRequire: Boolean,
): dynamic {
    val raw = js("{}")

    if (preLink != null) raw.pre = preLink.toJs(hasRequire)
    if (postLink != null) raw.post = postLink.toJs(hasRequire)

    return raw
}

File: integrations/scala/src/main/scala/angular/ts/Directive.scala
package angular.ts

import scala.scalajs.js

enum DirectiveRestrict(val value: String):
  case Attribute extends DirectiveRestrict("A")
  case Element extends DirectiveRestrict("E")
  case AttributeElement extends DirectiveRestrict("AE")
  case ElementAttribute extends DirectiveRestrict("EA")

final case class DirectivePrePost(
    pre: js.UndefOr[js.Function2[Scope, org.scalajs.dom.HTMLElement, Unit]] =
      js.undefined,
    post: js.UndefOr[js.Function2[Scope, org.scalajs.dom.HTMLElement, Unit]] =
      js.undefined,
):
  private[ts] def toJS: js.Object =
    JsObjectBuilder(
      "pre" -> pre,
      "post" -> post,
    )

type DirectiveFactoryFn = js.Function0[Directive | js.Function | js.Object]
type AnnotatedDirectiveFactory = js.Array[String | DirectiveFactoryFn]
type DirectiveFactory = DirectiveFactoryFn | AnnotatedDirectiveFactory

final case class Directive(
    link: js.UndefOr[js.Function2[Scope, org.scalajs.dom.HTMLElement, Unit]] =
      js.undefined,
    prePostLink: js.UndefOr[DirectivePrePost] = js.undefined,
    restrict: js.UndefOr[DirectiveRestrict] = js.undefined,
    controller: js.UndefOr[InjectableFactory[?]] = js.undefined,
    controllerAs: js.UndefOr[String] = js.undefined,
    bindToController: js.UndefOr[Boolean] = js.undefined,
    priority: js.UndefOr[Int] = js.undefined,
    terminal: js.UndefOr[Boolean] = js.undefined,
    replace: js.UndefOr[Boolean] = js.undefined,
    require: Map[String, String] = Map.empty,
    scope: js.UndefOr[Boolean] = js.undefined,
    template: js.UndefOr[String] = js.undefined,
    templateNamespace: js.UndefOr[String] = js.undefined,
    templateUrl: js.UndefOr[String] = js.undefined,
    transclude: js.UndefOr[Boolean] = js.undefined,
):
  private[ts] def factory: js.Function0[js.Object] =
    () =>
      val linkValue = prePostLink.fold(link.asInstanceOf[js.UndefOr[js.Any]])(
        value => value.toJS.asInstanceOf[js.Any],
      )

      JsObjectBuilder(
        "link" -> linkValue,
        "restrict" -> restrict.map(_.value),
        "controller" -> controller.map(_.annotated),
        "controllerAs" -> controllerAs,
        "bindToController" -> bindToController,
        "priority" -> priority,
        "terminal" -> terminal,
        "replace" -> replace,
        "require" -> JsObjectBuilder.stringMap(require),
        "scope" -> scope,
        "template" -> template,
        "templateNamespace" -> templateNamespace,
        "templateUrl" -> templateUrl,
        "transclude" -> transclude,
      )
